### PR TITLE
Refactor readers so Close() is called

### LINKED ISSUE #74: Need to Close ReadClosers used by importer pod
This is complicated by the fact that some of the intermediate Readers won't naturally call `Close`, eg. Tar.

### PR DESCRIPTION
**Pull request by jeffvance**
Fixes #74 
Restructure all Readers so that each reader's close is called within the same scope as the reader's creation.

@copejon @screeley44

### CODE REVIEW

**Comment by jeffvance** [comment]
For 2nd commit (ecadb612c) I get these 3 functional tests failures:
1.
```
    should unpack .tar.gz [It]
    /root/go/src/github.com/kubevirt/containerized-data-importer/test/datastream/datastream_test.go:100

    Expected error:
        <*errors.errorString | 0xc42003f870>: {
            s: "Copy: TarReader: reading tarfile header: archive/tar: invalid tar header\n\n",
        }
        Copy: TarReader: reading tarfile header: archive/tar: invalid tar header
```
2.
```
    should unpack .tar.xz [It]
    /root/go/src/github.com/kubevirt/containerized-data-importer/test/datastream/datastream_test.go:100

    Expected error:
        <*errors.errorString | 0xc420238130>: {
            s: "Copy: TarReader: reading tarfile header: archive/tar: invalid tar header\n\n",
        }
        Copy: TarReader: reading tarfile header: archive/tar: invalid tar header
```
3.
```
    should convert .qcow2 [It]
    /root/go/src/github.com/kubevirt/containerized-data-importer/test/datastream/datastream_test.go:100

    Expected error:
        <*errors.errorString | 0xc42003ee00>: {
            s: "Copy: copyImage: unable to stream data to file \"/tmp/tmp-/tinyCore.qcow2\": StreamDataToFile: create file error: open /tmp/tmp-/tinyCore.qcow2: no such file or directory\n",
        }
```

File: pkg/importer/dataStream.go [hunk 1]
```diff
@@ -148,53 +141,66 @@ func (d *dataStream) Copy(fn string) error {
...
 	}
 	qemu := image.MatchQcow2MagicNum(magicStr)
 
 	// Don't lose bytes read reading the magic number. MultiReader reads from each
 	// reader, in order, until the last reader returns eof.
 	multir := io.MultiReader(bytes.NewReader(magicStr), d.DataRdr)
 
-	// copy image file
-	out := common.IMPORTER_WRITE_PATH
+	// copy image file to hard-coded destination
```
Comment (copejon) [inline pkg/importer/dataStream.go]: > // copy image file to ~hard-coded~ destination

File: pkg/importer/dataStream.go [hunk 2]
```diff
@@ -116,6 +115,104 @@ func (d *dataStream) local() (io.ReadCloser, error) {
...
+func copyImage(r io.Reader, out string, qemu bool) error {
+	out = filepath.Clean(out)
+	glog.Infof("copyImage: copying image file to %q", out)
+	dest := out
+	if qemu {
+		// copy to tmp; the qemu conversion will write to final destination
+		dest = filepath.Join("/tmp", filepath.Base(out))
+		// make sure dest != out for qemu tmp location
+		if dest == out {
+			dest = filepath.Join("/tmp", "tmp41-" + filepath.Base(out))
```
Comment (jeffvance) [inline pkg/importer/dataStream.go]: use pseudo random name here

File: pkg/importer/dataStream.go [hunk 3]
```diff
@@ -116,6 +115,104 @@ func (d *dataStream) local() (io.ReadCloser, error) {
...
+} 
+
+// Copy the file using its Reader (r) to the passed-in destination (`out`).
+func copyImage(r io.Reader, out string, qemu bool) error {
+	out = filepath.Clean(out)
+	glog.Infof("copyImage: copying image file to %q", out)
+	dest := out
+	if qemu {
+		// copy to tmp; the qemu conversion will write to final destination
+		dest = filepath.Join("/tmp", filepath.Base(out))
```
Comment (jeffvance) [inline pkg/importer/dataStream.go]: add comment about /tmp maybe not having enough disk space

**Pull request merged**